### PR TITLE
[feature: #162727561]: In App Notification

### PR DESCRIPTION
**Pull request by XROLE**
#### What does this PR do?
- Handle inApp notification

#### Description of Task to be completed?
- [x] install pusher
- [x] add client directory as a frontend template to test inapp functionality
- [x] create follow model and migration
- [x] create methods to send notification to followers when new articles are created
- [x] create method to send notification to authors when there is comment on their ariticle
- [x] create notification model and migration
- [x] create endpoint to get all notification
- [x] create endpoint to edit the status of notification from unread to read
- [x] modify register and login route to return all notifications that belong to a user
- [x] write test for the functionalities

#### How should this be manually tested? 
- clone the repo
- run `npm install` to install new packages (Pusher) 
- run `npm install db:migrate` to create the new database
- Make sure; at least two users are created, a user has article in database and has atleas a follower
- fire up postmant
- navigate to PATCH `localhost:3000/api/v1/user/notifications?notificationId=<id of notification from notification table>` to change notification status from `unread` to `read`
- navigate to  GET `localhost:3000/api/v1/user/notifications` to get all notifications belonging to the user

#### What are the relevant pivotal tracker stories?
[#162727561](https://www.pivotaltracker.com/story/show/162727561)

### CODE REVIEW

**Comment by oreoluwade** [comment]
You have hound errors to fix and merge conflicts to resolve. Please do that.

File: app.js [hunk 1]
```diff
@@ -20,6 +20,7 @@ initPassport();
 
 app.use(cors());
 
+
```
Comment (oreoluwade) [inline app.js]: Please revert this change since you didn't do anything in this file.

File: client/client.html [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<!DOCTYPE html>
+<head>
+  <title>Pusher Test</title>
+  <link rel="stylesheet" href="style.css">
+  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.0/css/all.css" integrity="sha384-lZN37f5QGtY3VHgisS14W3ExzMWZxybE1SJSEsQp9S+oqd12jhcu+A56Ebc1zFSJ" crossorigin="anonymous">
+  <script src="https://js.pusher.com/4.3/pusher.min.js"></script>
+  <script>
+
+    // Enable pusher logging - don't include this in production
+    Pusher.logToConsole = true;
```
Comment (oreoluwade) [inline client/client.html]: Since you don't want this to be in production, there should be a conditional here:

`process.env.NODE_ENV === 'development'`

File: client/client.html [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+    var pusher = new Pusher('672d4a2dcf394b52247d', {
```
Comment (oreoluwade) [inline client/client.html]: The string you initialized the `Pusher` class with, where is it from?

File: client/client.html [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      cluster: 'eu',
+      forceTLS: true
+    });
+
+    var channel = pusher.subscribe('commentNotification');
+    channel.bind('11e928e9-9578-4b26-b052-76bbeb22d346', function(data) {
```
Comment (oreoluwade) [inline client/client.html]: This too? That variable string is from where?

File: client/main.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...JlZGNvbG9uaWFsIGp1c3QgY29tbWVudGVkIG9uIHlvdXIgYXJ0aWNsZTogSnVzdCBjYWxsIG1lIG1hcmsiLCJyZWRjb2xvbmlhbCBqdXN0IGNvbW1lbnRlZCBvbiB5b3VyIGFydGljbGU6IEp1c3QgY2FsbCBtZSBtYXJrIiwicmVkY29sb25pYWwganVzdCBjb21tZW50ZWQgb24geW91ciBhcnRpY2xlOiBKdXN0IGNhbGwgbWUgbWFyayIsInJlZGNvbG9uaWFsIGp1c3QgY29tbWVudGVkIG9uIHlvdXIgYXJ0aWNsZTogSnVzdCBjYWxsIG1lIG1hcmsiLCJyZWRjb2xvbmlhbCBqdXN0IGNvbW1lbnRlZCBvbiB5b3VyIGFydGljbGU6IEp1c3QgY2FsbCBtZSBtYXJrIiwicmVkY29sb25pYWwganVzdCBjb21tZW50ZWQgb24geW91ciBhcnRpY2xlOiBKdXN0IGNhbGwgbWUgbWFyayIsInJlZGNvbG9uaWFsIGp1c3QgY29tbWVudGVkIG9uIHlvdXIgYXJ0aWNsZTogSnVzdCBjYWxsIG1lIG1hcmsiLCJyZWRjb2xvbmlhbCBqdXN0IGNvbW1lbnRlZCBvbiB5b3VyIGFydGljbGU6IEp1c3QgY2FsbCBtZSBtYXJrIiwicmVkY29sb25pYWwganVzdCBjb21tZW50ZWQgb24geW91ciBhcnRpY2xlOiBKdXN0IGNhbGwgbWUgbWFyayIsInJlZGNvbG9uaWFsIGp1c3QgY29tbWVudGVkIG9uIHlvdXIgYXJ0aWNsZTogSnVzdCBjYWxsIG1lIG1hcmsiLCJyZWRjb2xvbmlhbCBqdXN0IGNvbW1lbnRlZCBvbiB5b3VyIGFydGljbGU6IEp1c3QgY2FsbCBtZSBtYXJrIiwicmVkY29sb25pYWwganVzdCBjb21tZW50ZWQgb24geW91ciBhcnRpY2xlOiBKdXN0IGNhbGwgbWUgbWFyayIsInJlZGNvbG9uaWFsIGp1c3QgY29tbWVudGVkIG9uIHlvdXIgYXJ0aWNsZTogSnVzdCBjYWxsIG1lIG1hcmsiXSwiaWF0IjoxNTQ4ODM1MDMyLCJleHAiOjE1NDkwOTQyMzJ9.ywYtbr7Lb4Xl2H97z3PoRCoJFqCPi7KE6jYiA6KK-Cw';
```
[the remark this answers is not in the document]
Comment (oreoluwade) [reply client/main.js]: You should use an eslint-disable since this token appears to be important. Otherwise, this should be in the `.env`

File: client/main.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+// eslint-disable-next-line require-jsdoc
+function fetchNotification() {
```
[the remark this answers is not in the document]
Comment (oreoluwade) [reply client/main.js]: Why wasn't this function used?

[the remark this answers is not in the document]
Comment (XROLE) [reply client/main.js]: It is used in the index.html file

File: client/main.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (XROLE) [reply client/main.js]: Ok. implementing the feedback

File: app.js (hunk 1 above)
Comment (XROLE) [reply app.js]: Ok. Noted

File: client/client.html (hunk 3 above)
Comment (XROLE) [reply client/client.html]: It is the user.id. By default, it is meant to come from decoded token in frontend.

File: client/client.html (hunk 4 above)
Comment (XROLE) [reply client/client.html]: The variables are user.id. Because when sending notification I will need a unique identifier, I chose userId to be the unique identifier. Though it is hardcoded because the client directory does nothing more than to test the inApp functionality. But, during the implementation, the id will come from `jwt.decoded.id` as against hard coding it.

File: client/client.html (hunk 3 above)
Comment (oreoluwade) [reply client/client.html]: So whose ID is the one currently hardcoded here?

File: client/main.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (oreoluwade) [reply client/main.js]: Okay. Put an `export` before the function definition. That should get rid of the eslint flagging

File: client/main.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    method: 'GET',
+    headers: {
+      Accept: 'application/json, text/plain, */*',
+      Authorization: token,
+      'Content-type': 'application/json'
+    }
+  })
+    .then(response => response.json())
+    .then((myJson) => {
+      const notificationIcon = document.getElementById('number');
```
[the remark this answers is not in the document]
Comment (oreoluwade) [reply client/main.js]: For this, you'd need to add a configuration option to the `.eslintrc` to allow the globally available `document` object. You can google how to allow `document` in a js file that eslint is flagging.

File: client/client.html (hunk 3 above)
Comment (XROLE) [reply client/client.html]: It was the user.id from user table in my database.

File: client/client.html (hunk 2 above)
Comment (XROLE) [reply client/client.html]: Ok. Noted

File: client/main.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (XROLE) [reply client/main.js]: ok. Thanks

File: services/notifier.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import models from '../models';
+import pusherConfig from './pusherConfig';
+
+const { Article, User, Notification } = models;
+
+export default {
+  async notifyInApp(articleId, commenterId) {
```
Comment (oreoluwade) [inline services/notifier.js]: Please rename to `newCommentNotification`

File: services/notifier.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    const commenterName = commenter.username;
+
+    const body = `${commenterName} just commented on your article: ${title}`;
+
+    await Notification.create({ userId, body, status: 'unread' });
+
+    pusherConfig.trigger('notification', userId, {});
+  },
+
+  async notifyFollowers(followerId, authorId, articleTitle) {
```
Comment (oreoluwade) [inline services/notifier.js]: Rename this to `notifyFollowersOfNewArticle`

Comment (XROLE) [reply services/notifier.js]: Ok. right away

File: services/notifier.js (hunk 8 above)
Comment (XROLE) [reply services/notifier.js]: Done

**Review by oreoluwade — APPROVED** [review]
